prusti-specs/src/specifications/preparser.rs

```diff
@@ -289,557 +58,547 @@ impl Parse for SpecEntArgs {
-        // detect possibly ambiguous input
-        if self.input.contains_both_and_or() {
-            return Err(self.error_ambiguous_expression());
-        }
-
-        // preparse the input into atomic Prusti assertions
-        while !self.input.is_empty() {
-            if self.input.check_and_consume_operator("&&") {
-                if let Err(err) = self.resolve_and() {
-                    return Err(err);
-                }
-            }
-            else if self.input.check_and_consume_operator("==>") {
-                return self.resolve_implies();
+    /// does the given operator appear in the stream anywhere
+    fn contains_operator_recursive(&self, stream: &VecDeque<TokenTree>, operator: &str) -> Option<Span> {
+        let mut stream = stream.clone();
+        while !stream.is_empty() {
+            if self.peek_operator_stream(&stream, operator) {
+                return Some(self.operator_span(&stream, operator));
             }
-            else if self.input.check_and_consume_keyword("forall") {
-                if let Err(err) = self.resolve_forall() {
-                    return Err(err);
+            if let Some(TokenTree::Group(group)) = stream.front() {
+                let nested_stream: VecDeque<TokenTree> = group.stream().into_iter().collect();
+                if let Some(span) = self.contains_operator_recursive(&nested_stream, operator) {
+                    return Some(span);
                 }
             }
-            else if self.input.check_and_consume_operator("|=") {
-                if let Err(err) = self.resolve_spec_ent() {
-                    return Err(err);
-                }
+            stream.pop_front();
+        }
+        None
+    }
+    /// get the span of the peeked operator
+    fn operator_span(&self, stream: &VecDeque<TokenTree>, operator: &str) -> Span {
+        let mut span: Option<Span> = None;
+        for (i, _) in operator.char_indices() {
+            let character = stream.get(i).unwrap();
+            if let Some(maybe_span) = span {
+                span = maybe_span.join(character.span());
+            } else {
+                span = Some(character.span());
             }
-            else if let Some(group) = self.input.check_and_consume_parenthesized_block() {
-                if let Err(err) = self.resolve_parenthesized_block(group) {
-                    return Err(err);
+        }
+        span.unwrap()
+    }
+    /// Check if there is a subexpression (parenthesized or separated by `==>`)
+    /// that contains both `&&` and `||`. If yes, set the span to include both
+    /// of those operators and everything in between them. This detects
+    /// potentially ambiguous subexpressions.
+    fn contains_both_and_or(&self, stream: &VecDeque<TokenTree>) -> Option<Span> {
+        let mut contains_and = false;
+        let mut contains_or = false;
+        let mut and_span: Option<Span> = None;
+        let mut or_span: Option<Span> = None;
```

Comment:
`contains_and == and_span.is_some()`, same for `or`. Remove the booleans.

Reply:
Addressed by c77b6d0b1638c36b81eddb1e87af37c3f1fff212